src/SignalR/server/Core/src/Internal/AsyncEnumeratorAdapters.cs

```diff
@@ -11,74 +12,89 @@ namespace Microsoft.AspNetCore.SignalR.Internal
     // True-internal because this is a weird and tricky class to use :)
     internal static class AsyncEnumeratorAdapters
     {
-        public static IAsyncEnumerator<object> GetAsyncEnumerator<T>(ChannelReader<T> channel, CancellationToken cancellationToken = default(CancellationToken))
+        public static IAsyncEnumerator<object> GetAsyncEnumeratorFromAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken = default(CancellationToken))
         {
-            // Nothing to dispose when we finish enumerating in this case.
-            return new AsyncEnumerator<T>(channel, cancellationToken, disposable: null);
+            var enumerator = asyncEnumerable.GetAsyncEnumerator(cancellationToken);
+            return enumerator as IAsyncEnumerator<object> ?? new BoxedAsyncEnumerator<T>(enumerator);
+        }
+
+        public static IAsyncEnumerator<object> GetAsyncEnumeratorFromChannel<T>(ChannelReader<T> channel, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return new ChannelAsyncEnumerator<T>(channel, cancellationToken);
+        }
+
+        /// <summary>Converts an IAsyncEnumerator of T to an IAsyncEnumerator of object.</summary>
+        private class BoxedAsyncEnumerator<T> : IAsyncEnumerator<object>
+        {
+            private IAsyncEnumerator<T> _asyncEnumerator;
+
+            public BoxedAsyncEnumerator(IAsyncEnumerator<T> asyncEnumerator)
+            {
+                _asyncEnumerator = asyncEnumerator;
+            }
+
+            public object Current => _asyncEnumerator.Current;
+
+            public ValueTask<bool> MoveNextAsync()
+            {
+                return _asyncEnumerator.MoveNextAsync();
+            }
+
+            public ValueTask DisposeAsync()
+            {
+                return _asyncEnumerator.DisposeAsync();
+            }
         }
 
         /// <summary>Provides an async enumerator for the data in a channel.</summary>
-        internal class AsyncEnumerator<T> : IAsyncEnumerator<object>, IDisposable
+        private class ChannelAsyncEnumerator<T> : IAsyncEnumerator<object>
```

Comment:
This won't be required in .NET Core 3.0. ChannelReader implements IAsyncEnumerable

Reply:
AFAIK the work hasn't been done for this yet. https://github.com/dotnet/corefx/issues/32742

Reply:
It's done now FWIW